irohad/synchronizer/impl/synchronizer_impl.cpp

```diff
@@ -176,28 +191,28 @@ namespace iroha {
 
       assert(height_diff > 0);
 
-      auto ledger_state = downloadMissingBlocks(
+      auto commit_result = downloadAndCommitMissingBlocks(
           *top_block_height, required_height, msg.public_keys);
 
-      shared_model::interface::types::HeightType new_height;
-      // TODO 11.04.2019 mboldyrev IR-442 use storage commit status type
-      BOOST_ASSERT_MSG(ledger_state, "Commit failed!");
-      if (ledger_state) {
-        new_height = (*ledger_state)->top_block_info.height;
-      } else {
-        log_->critical(
-            "Synchronization cancelled because "
-            "downloaded blocks commit failed!");
-        return;
-      }
-
-      const bool higher_than_expected = new_height > required_height;
-      notifier_.get_subscriber().on_next(SynchronizationEvent{
-          higher_than_expected ? SynchronizationOutcomeType::kCommit
-                               : alternative_outcome,
-          higher_than_expected ? consensus::Round{new_height, 0} : msg.round,
-          ledger_state ? std::move(*ledger_state)
-                       : std::unique_ptr<LedgerState>{}});
+      commit_result.match(
+          // TODO make it auto after #2226
+          [this, required_height, alternative_outcome, &msg](
+              ametsuchi::CommitResult::ValueType &value) {
+            auto &ledger_state = value.value;
+            assert(ledger_state);
```

Comment:
I guess in release builds this assertion will be thrown off here. My suggestion is to add a log message similar to the old code.

Reply:
Hello me 19 days ago^^

Reply:
this situation is unrecoverable. I should have written it as we discussed then, that such assertions are ok to be checked in tests and left as-is in production builds.

Reply:
Is this an assertion for `nullptr`? If yes, it can be removed, since it is a convention that there are no `nullptr`s.